logic/resolvable/Retrievable.java

```diff
@@ -35,4 +76,196 @@ public Retrievable asRetrievable() {
     public boolean isRetrievable() {
         return true;
     }
-}
\ No newline at end of file
+
+    public static class Extractor {
+        private final Conjunction conjunction;
+        private final Set<Concludable<?>> concludables;
+
+        private final Map<Variable, Variable> variablesMap = new HashMap<>(); // Map from variables in original conjunction to new subConjunction
+        private final Set<Constraint> visitedConstraints = new HashSet<>();
+        private final Set<Set<Variable>> connectedSubgraphs = new HashSet<>();
+        private Set<Variable> connectedSubgraph;
+
+
+        public Extractor(Conjunction conjunction, Set<Concludable<?>> concludables) {
+            this.conjunction = conjunction;
+            this.concludables = concludables;
+        }
+
+        public static Extractor from(Conjunction conjunction, Set<Concludable<?>> concludables) {
+            return new Extractor(conjunction, concludables);
+        }
+
+        public Set<Retrievable> extract() {
+            concludables.forEach(concludable -> visitedConstraints.addAll(concludable.coreConstraints()));
+            conjunction.variables().stream().filter(var -> var.identifier().reference().isName()).forEach(var -> {
+                if (variablesMap.get(var) == null) {
+                    connectedSubgraph = new HashSet<>();
+                    if (var.isThing()) addThingVariable(var.asThing());
+                    else if (var.isType()) addTypeVariable(var.asType());
+                    else throw GraknException.of(ILLEGAL_STATE);
+                    connectedSubgraphs.add(connectedSubgraph);
+                }
+            });
+            return connectedSubgraphs.stream()
+                    .filter(subGraph -> !(subGraph.size() == 1 && subGraph.iterator().next().constraints().size() == 0))
+                    .map(subGraph -> new Retrievable(new Conjunction(subGraph, set()))).collect(Collectors.toSet());
+        }
+
+        private void addToConnected(Variable copy) {
+            connectedSubgraph.add(copy);
+        }
+
+        private TypeVariable addTypeVariable(TypeVariable variable) {
```

Comment:
I believe this is the right time to use `register` - `registerTypeVariable()` and `registerThingVariable()`